plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePartitionsTable.java

```diff
@@ -0,0 +1,461 @@
+    {
+        if (partitionColumns.isEmpty()) {
+            return new EmptyPageSource();
+        }
+
+        return new FixedPageSource(buildPages(session));
+    }
+
+    private List<Page> buildPages(ConnectorSession session)
+    {
+        PageListBuilder pageListBuilder = PageListBuilder.forTable(tableMetadata);
+        MetadataEntry metadataEntry = transactionLogAccess.getMetadataEntry(session, tableSnapshot);
+        ProtocolEntry protocolEntry = transactionLogAccess.getProtocolEntry(session, tableSnapshot);
+
+        try (Stream<AddFileEntry> activeFiles = transactionLogAccess.loadActiveFiles(session, tableSnapshot, metadataEntry, protocolEntry, TupleDomain.all(), alwaysTrue())) {
+            for (Map.Entry<Map<String, Optional<String>>, DeltaLakePartitionStatistics> partitionEntry : getStatisticsByPartition(activeFiles).entrySet()) {
+                Map<String, Optional<String>> partitionValue = partitionEntry.getKey();
+                DeltaLakePartitionStatistics deltaLakePartitionStatistics = partitionEntry.getValue();
+
+                RowType partitionValuesRowType = RowType.from(partitionFields);
+                SqlRow partitionValuesRow = buildRowValue(partitionValuesRowType, fields -> {
+                    for (int i = 0; i < partitionColumns.size(); i++) {
+                        DeltaLakeColumnHandle column = partitionColumns.get(i);
+                        Type type = column.getType();
+                        Optional<String> value = partitionValue.get(column.getBasePhysicalColumnName());
+                        Object deserializedPartitionValue = deserializePartitionValue(column, value);
+                        writeNativeValue(type, fields.get(i), deserializedPartitionValue);
+                    }
+                });
+
+                pageListBuilder.beginRow();
+
+                pageListBuilder.appendNativeValue(partitionValuesRowType, partitionValuesRow);
+                pageListBuilder.appendBigint(deltaLakePartitionStatistics.fileCount());
+                pageListBuilder.appendBigint(deltaLakePartitionStatistics.size());
+
+                dataColumnType.ifPresent(dataColumnType -> {
+                    try {
+                        SqlRow dataColumnRow = buildRowValue(dataColumnType, fields -> {
+                            for (int i = 0; i < columnMetricTypes.size(); i++) {
+                                String fieldName = regularColumns.get(i).getBaseColumnName();
+                                Object min = deltaLakePartitionStatistics.minValues().getOrDefault(fieldName, null);
+                                Object max = deltaLakePartitionStatistics.maxValues().getOrDefault(fieldName, null);
+                                Long nullCount = deltaLakePartitionStatistics.nullCounts().getOrDefault(fieldName, null);
+                                if (min == null && max == null && nullCount == null) {
+                                    throw new MissingColumnMetricsException();
```

Comment:
We don't need an exception here.
Simply set a flag value to `false` to point out that the partitions are missing and do the handling after the `for`

Reply:
Don't think this is possible, see following snippet:

```
dataColumnType.ifPresent(dataColumnType -> {
                    boolean appendDataColumn = true;
                    SqlRow dataColumnRow = buildRowValue(dataColumnType, fields -> {
                        for (int i = 0; i < columnMetricTypes.size(); i++) {
                            String fieldName = regularColumns.get(i).getBaseColumnName();
                            Object min = deltaLakePartitionStatistics.minValues().getOrDefault(fieldName, null);
                            Object max = deltaLakePartitionStatistics.maxValues().getOrDefault(fieldName, null);
                            Long nullCount = deltaLakePartitionStatistics.nullCounts().getOrDefault(fieldName, null);
                            if (min == null && max == null && nullCount == null) {
                                appendDataColumn = false;
                                break;
                            }

                            RowType columnMetricType = columnMetricTypes.get(i);
                            columnMetricType.writeObject(fields.get(i), getColumnMetricBlock(columnMetricType, min, max, nullCount));
                        }
                    });

                    if (appendDataColumn) {
                        pageListBuilder.appendNativeValue(dataColumnType, dataColumnRow);
                    }
                    else {
                        pageListBuilder.appendNull();
                    }
                });
```

yields the following error: `DeltaLakePartitionsTable.java:[197,33] local variables referenced from a lambda expression must be final or effectively final`. Line 197 is `appendDataColumn = false;`.

Reply:
Use `AtomicReference<Boolean>`